core/src/consensus.rs

```diff
@@ -606,6 +703,203 @@ impl Tower {
         bank_weight
     }
 
+    fn voted_slots(&self) -> Vec<Slot> {
+        self.lockouts
+            .votes
+            .iter()
+            .map(|lockout| lockout.slot)
+            .collect()
+    }
+
+    pub fn is_stray_last_vote(&self) -> bool {
+        if let Some(last_voted_slot) = self.last_voted_slot() {
+            if let Some(max_stray_slot) = self.stray_restored_slots.iter().next_back() {
+                return *max_stray_slot == last_voted_slot;
+            }
+        }
+
+        false
+    }
+
+    fn stray_restored_ancestors(&self) -> HashSet<Slot> {
+        let mut ancestors: HashSet<_> = self.stray_restored_slots.clone().into_iter().collect();
+        if let Some(last_voted_slot) = self.last_voted_slot() {
+            assert!(ancestors.contains(&last_voted_slot));
+            ancestors.remove(&last_voted_slot);
+        }
+        ancestors
+    }
+
+    // The tower root can be older/newer if the validator booted from a newer/older snapshot, so
+    // tower lockouts may need adjustment
+    pub fn adjust_lockouts_after_replay(
+        self,
+        replayed_root_slot: Slot,
+        slot_history: &SlotHistory,
+    ) -> Result<Self> {
+        info!(
+            "adjusting lockouts (after replay up to {}): {:?}",
+            replayed_root_slot,
+            self.voted_slots()
+        );
+
+        assert_eq!(slot_history.check(replayed_root_slot), Check::Found);
+        // reconcile_blockstore_roots_with_tower() should already have aligned these.
+        assert!(
+            self.root().is_none() || self.root().unwrap() <= replayed_root_slot,
```

Comment:
I may be missing something, but will this `self.root().unwrap() <= replayed_root_slot` assertion fail if we jump ahead and start from a snapshot in the future?

For instance, say the chain of slots is `1, 2, 3, 4, 5`, our latest root is `2`, but then we go offline and start from a snapshot for `5` that we had not previously replayed. Then `replayed_root_slot` will be `5` at startup, but `self.root() == 2`?

Reply:
>  Then `replayed_root_slot` will be `5` at startup, but `self.root() == 2`?

Well, assertion shouldn't fail. because `self.root().unwrap() <= replayed_root_slot` could be substituted to `2 <= 5`, then `true`. So the assertion evaluates to `true` without causing `panic!`.

So, I think this assertion is fine as-is.

This asserts `replayed_root_slot` must always be greater than or equal to `self.root()` as a sanity check because we're always setting all the ancestors of `self.root()` as root at `reconcile_blockstore_roots_with_tower()` before replaying. So after replaying, `replayed_root_slot` must always proceed to `self.root()` at the very least.

Reply:
Heh, I misread the `<=` as `>=`, sorry!